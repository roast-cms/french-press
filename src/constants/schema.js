import {Block} from "slate"

/**
 * A set of rules that transform document structure and keep it normalized to a defined format.
 * @constant schema
 * @property {Array} nodes Acceptable nodes within editor document.
 * @property {Object} last Defines at least one empty paragraph block that follows a void block (such as picture); this is required to ensure that the user can continue adding content without additional effort below uploaded images (otherwise they will be forced to move the image up to free up a trailing paragraph space).
 */

export const SCHEMA = {
  document: {
    nodes: [
      {
        types: [
          "paragraph",
          "heading",
          "divider",
          "quote",
          "image",
          "docket",
          "link",
        ],
      },
    ],
    last: {types: ["paragraph", "quote"]},
    normalize: (change, reason, {node}) => {
      switch (reason) {
        case "last_child_type_invalid": {
          const paragraph = Block.create("paragraph")
          return change.insertNodeByKey(node.key, node.nodes.size, paragraph)
        }
        default:
          return null
      }

    },
  },
  blocks: {
    link: {
      nodes: [{objects: ["text"]}],
    },
    divider: {
      isVoid: true,
    },
    image: {
      isVoid: true,
      data: {
        src: value => value,
      },
      parent: [{object: "document"}],
      normalize: (change, {code, node}) => {
        if (code === "parent_object_invalid") {
          return change.unwrapBlockByKey(node.key)
        } else return null
      },
    },
    docket: {
      isVoid: true,
    },
  },
  inlines: {
    quote: {
      nodes: [{types: ["text"]}],
    },
    paragraph: {
      nodes: [{types: ["text", "link"]}],
    },
    heading: {
      nodes: [{types: ["text"]}],
    },
  },
}
