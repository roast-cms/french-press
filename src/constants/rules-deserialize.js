/**
 * A set of rules that defines how to transpile HTML itno slate document nodes and marks (and back).
 * @constant rules
 */
import isUrl from "is-url"

import {BLOCK_TAGS, MARK_TAGS} from "./defaults"

/**
 * Flattens HTML into plain text.
 * @function squish
 * @return {String}
 */
const squish = el => {
  el.innerHTML = el.innerText || el.textContent
  return el
}


export const RULES_DESERIALIZE = [
  {
    deserialize(el, next) {
      const block = BLOCK_TAGS[el.tagName.toLowerCase()]
      if (!block) return
      switch (block) {
        case "paragraph": {
          return {
            object: "block",
            type: "paragraph",
            nodes: next(el.childNodes),
          }
        }
        case "quote": {
          return {
            object: "block",
            type: "quote",
            nodes: next(el.childNodes),
          }
        }
        case "heading": {
          return {
            object: "block",
            type: "heading",
            nodes: next(el.childNodes),
          }
        }
        case "image": {
          const src = el.getAttribute("src") || el.getAttribute("srcset")
          if (!isUrl(src)) return
          return {
            object: "block",
            type: "image",
            isVoid: true,
            data: {src},
          }
        }
        case "link": {
          return {
            object: "inline",
            type: "link",
            data: {
              href: el.getAttribute("href"),
            },
            nodes: next(el.childNodes),
          }
        }
        default:
          return {
            object: "block",
            type: "paragraph",
            nodes: next(el.childNodes),
          }
      }
    },
  },
  {
    deserialize(el, next) {
      const mark = MARK_TAGS[el.tagName]
      if (!mark) return
      return {
        object: "mark",
        type: mark,
        nodes: next(el.childNodes),
      }
    },
  },
]
