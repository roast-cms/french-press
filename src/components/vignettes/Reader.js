import React from "react"

import {DEFAULT_EDITOR_STATE} from "../../constants/defaults"
import {
  RULES_SERIALIZE,
  rulesSerializeWithProps,
} from "../../constants/rules-serialize"

const rules = props => [
  ...RULES_SERIALIZE,
  ...rulesSerializeWithProps(props),
  {
    serialize(node, children) {
      if (node.object === "string") {
        return children.split("\n").reduce((array, text, i) => {
          if (i !== 0) array.push(<br key={i} />)
          array.push(text)
          return array
        }, [])
      }
    },
  },
]

export const addRootSerialNumbers = nodes => {
  nodes.forEach((node, iterator) => {
    node.serial = iterator
  })
  return nodes
}
let nodes = []
export const nodeFactory = props => {
  const {document} = props.value || DEFAULT_EDITOR_STATE
  nodes = addRootSerialNumbers(document.nodes)
  const elements = nodes.map(node => serializeNode(node, props)).filter(Boolean)
  return elements
}
export const serializeNode = (node, props) => {
  if (node.object === "text") {
    const {leaves} = node
    return leaves.map(leaf => serializeLeaf(leaf, props))
  }
  const children = node.nodes.map(serializeNode)
  for (const rule of rules(props)) {
    if (!rule.serialize) continue
    const ret = rule.serialize(node, children)
    if (ret === null) return
    if (ret) return addKey(ret)
  }
  throw new Error(`No serializer defined for node of type "${node.type}".`)
}
export const serializeLeaf = (leaf, props) => {
  const string = {object: "string", text: leaf.text}
  const text = serializeString(string, props)
  if (!leaf.marks) return leaf.text

  return leaf.marks.reduce((children, mark) => {
    for (const rule of rules(props)) {
      if (!rule.serialize) continue
      const ret = rule.serialize(mark, children)
      if (ret === null) return null
      if (ret) return addKey(ret)
    }
    throw new Error(`No serializer defined for mark of type "${mark.type}".`)
  }, text)
}
export const serializeString = (string, props) => {
  for (const rule of rules(props)) {
    if (!rule.serialize) continue
    const ret = rule.serialize(string, string.text)
    if (ret) return ret
  }
}
let key = 0
export const addKey = element => {
  const thisKey = key++
  return React.cloneElement(element, {
    key: thisKey,
    node: element.props.node
      ? {
          ...element.props.node,
          get: object => {
            return object === "key" && thisKey
          },
        }
      : null,
    parent: {
      getNextBlock: () => {
        const node = nodes.filter(object => {
          return object.serial === element.props.node.serial + 1
        })[0]
        const nodeFunction = {
          get: object =>
            object === "data" && {
              get: object => (node.data ? node.data[object] : undefined),
            },
        }
        return nodeFunction
      },
    },
  })
}

export default props => nodeFactory(props)
