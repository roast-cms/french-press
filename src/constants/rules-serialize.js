import React from "react"

import {addKey} from "../components/vignettes/Reader"
import Link from "../components/controls/Link"

export const rulesSerializeWithProps = props => [
  {
    serialize(node, children) {
      const element = node.type
      switch (element) {
        case "link": {
          const href = node.data.href
          return addKey(
            <Link
              to={href}
              domain={props.options ? props.options.domain : null}
            >
              {children}
            </Link>
          )
        }
        case "image": {
          if (props.components && props.components.Picture) {
            const Picture = props.editor
              ? props.editor.props.components.Picture
              : props.components.Picture
            return addKey(
              <Picture
                editor={{value: {isSelected: false}}}
                node={{
                  data: {
                    get: object => node.data[object],
                  },
                  serial: node.serial,
                }}
                readOnly={true}
              />
            )
          } else return null
        }
        default:
          return addKey(<span>{children}</span>)
      }
    },
  },
]
export const RULES_SERIALIZE = [
  {
    serialize(node, children) {
      const element = node.type
      switch (element) {
        case "paragraph": {
          return addKey(<p>{children}</p>)
        }
        case "quote": {
          return addKey(
            <div style={{clear: "both"}}>
              <blockquote>{children}</blockquote>
            </div>
          )
        }
        case "heading": {
          return addKey(<h3>{children}</h3>)
        }
        case "divider": {
          return addKey(<hr />)
        }
        case "italic": {
          return addKey(<em>{children}</em>)
        }
        case "bold": {
          return addKey(<strong>{children}</strong>)
        }
      }
    },
  },
]
