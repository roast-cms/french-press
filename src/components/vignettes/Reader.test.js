import "jest-styled-components"

import React from "react"

import {shallow} from "enzyme"

import {DEFAULT_EDITOR_STATE} from "../../constants"
import {Reader} from "./Reader"

test("Render Reader without crashing, matches snapshot", () => {
  const element = shallow(<Reader value={DEFAULT_EDITOR_STATE} />)
  expect(element).toMatchSnapshot()
})
