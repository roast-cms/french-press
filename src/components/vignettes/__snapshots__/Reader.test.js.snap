// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Reader without crashing, matches snapshot 1`] = `ShallowWrapper {}`;

exports[`addKey creates React elements that match snapshot. 1`] = `
<type
  node={null}
/>
`;

exports[`serializeLeaf maps leaves into marks components with keys 1`] = `
<em
  node={null}
>
  <span
    node={null}
  >
    This is the first post to go onto this website.
  </span>
</em>
`;

exports[`serializeString returns string, wrapped in consistent components. 1`] = `
<span
  node={null}
>
  This is the first post to go onto this website.
</span>
`;
